doc/src/explanation/best-practices.md

```diff
@@ -0,0 +1,1329 @@
+  the object](best-practices-args-invariants) (e.g., something like
+  `YourObject(*instance.args)` should recreate `instance`).
+
+  Additionally, it should be mentioned that it is not a good idea to subclass
+  `Symbol` if you plan to store anything extra in `args`. `Symbol` is designed
+  around having no `args`. You are better off subclassing `Function` (see
+  [](custom-functions)) or `Expr` directly. If you simply want to have two
+  symbols that are distinct from one another, the best approach is often just
+  to give them different names. If you are concerned about how they are
+  printed, you can replace them with a more canonical name when it comes time
+  to print things, or use a [custom printer](module-printing).
+
+- **Store the data about the object separately.** This is the best approach if
+  the extra data is not directly related to an object's mathematical
+  properties.
+
+  Remember that SymPy objects are hashable, so they can easily be used as
+  dictionary keys. So maintaining a separate dictionary of `{object:
+  extra_data}` pairs is straightforward.
+
+  Note that some SymPy APIs already allow redefining how they operate on
+  objects separately from the objects themselves. A big example of this is the
+  {term}`printers <printing>`, which allow defining [custom
+  printers](module-printing) that change how any SymPy object is printed
+  without modifying those object themselves. Functions like {func}`~.lambdify`
+  and {func}`~.init_printing` allow passing in a custom printer.
+
+- **Represent the attribute using different subclasses.** This is often a good
+  idea if there are only a few possible values for the attribute (e.g., a
+  boolean flag). Code duplication can be avoided by using a common superclass.
+
+- **If the data you want to store is a Python function**, it's best to just
+  use as a method on the class. In many cases, the method may already fit into
+  one of the [existing set of overridable SymPy methods](custom-functions). If
+  you want to define how a function evaluates itself numerically, you can use
+  {func}`~.implemented_function`.
+
+- **Represent the information using by modifying the object's `func`.** This
+  solution is much more complicated than the others, and should only be used
+  when it is necessary. In some extreme cases, it is not possible to represent
+  every mathematical aspect of an object using `args` alone. This can happen,
+  for example, because of the limitation that [`args` should only contain
+  `Basic` instances](best-practices-args-invariants). It is still possible to
+  create custom SymPy objects in these situations by using a custom
+  {term}`func` that is different from `type(expr)` (in this case, you would
+  override `__eq__` on the `func` [rather than on the
+  class](best-practices-eq)).
+
+  However, this sort of situation is rare.
+
+(best-practices-eq)=
+### Don't Overwrite `__eq__`
```

Comment:
I don't know if we want to mention this here but related to "avoid extra attributes" and "don't override `__eq__`" is the question of what to do when you *do* need to do these things.

It is rare that these things are needed but there are examples e.g. `Poly`. Storing the `args` of a Poly as Expr would be very inefficient so it is not done and also the domain argument is not representable as `Basic` (possibly what this shows is that `Poly` should not be `Basic` at all). Likewise `ImmutableMatrix` has a `DomainMatrix` as its internal representation and it would be inefficient to always store the `args` in that case.

A more common situation where you do need to attach additional data not in the `args` is for atoms.

The question then is what you *should* do e.g. to implement an atom or some other Basic subclass whose constructor arguments will *not* end up in the args or at least will not be stored explicitly in `_args`.

1. Don't override `__eq__` or `__hash__`: use `_hashable_content` instead.
2. Override `__new__` but *don't* pass the args to `super().__new__`.
3. Implement `__getnewargs__` (`__getnewargs_ex__` if keyword arguments are needed) for pickling etc.
4. If not an `Atom` then override `.args` as a property that will generate args on demand (and ensure that the `func-args` invariant holds).

Reply:
These are somewhat advanced things. I'm not even sure myself what the proper way to do some of those things would be (is it ever actually a good idea to override `_hashable_content`?). In my experience, most people who are trying to override `__eq__` are completely misguided and shouldn't be doing it at all. They are just trying to do something that goes against the SymPy norms.

I did mention overriding `func` as an advanced escape hatch. Maybe I can briefly mention these other things, but if anything, I feel like this just gets to the core of why this is a bad idea: there are lot of things here and it's very easy to get it wrong.